lightning/src/ln/channelmanager.rs

```diff
@@ -12321,6 +12324,41 @@ where
 		)
 	}
 
+	/// Pays for an [`Offer`] which was built by resolving a human readable name. It is otherwise
+	/// identical to [`Self::pay_for_offer`].
+	pub fn pay_for_offer_from_hrn(
+		&self, offer: &OfferFromHrn, amount_msats: u64, payment_id: PaymentId,
+		optional_params: OptionalOfferPaymentParams,
+	) -> Result<(), Bolt12SemanticError> {
+		let create_pending_payment_fn = |invoice_request: &InvoiceRequest, nonce| {
+			let expiration = StaleExpiration::TimerTicks(1);
+			let retryable_invoice_request = RetryableInvoiceRequest {
+				invoice_request: invoice_request.clone(),
+				nonce,
+				needs_retry: true,
+			};
+			self.pending_outbound_payments
+				.add_new_awaiting_invoice(
+					payment_id,
+					expiration,
+					optional_params.retry_strategy,
+					optional_params.route_params_config,
+					Some(retryable_invoice_request),
+				)
+				.map_err(|_| Bolt12SemanticError::DuplicatePaymentId)
+		};
```

Comment:
A little different from my earlier suggestions, but to avoid duplicating this, we could consider:
-  implementing `AsRef<Offer>` for `OfferFromHrn`
- adding an `ToHumanReadableName` trait implemented by both `Offer` and `OfferFromHrn` with a method returning `Option<HumanReadableName>`
- using a single method taking `O: AsRef<Offer> + ToHumanReadableName`

That way `OfferFromHrn` could be private to help avoid misuse. Though maybe explicit methods for each case would be more intuitive?